tests/Highlighter.py

```diff
@@ -0,0 +1,358 @@
+            if i%2 == 0: code.append(line)
+            else: expected.append(line)
+
+        code = "\n".join(code)
+        expected = "\n".join(expected)
+
+        self.assert_highlighted_chr(code, expected, level, lang)
+
+
+    def assert_highlighted(self, code_col, level, lang="en"):
+        code, expected = self.converts(code_col)
+        self.assert_highlighted_chr(code, expected, level, lang)
+
+
+    def assert_highlighted_multi_line(self, *args, level, lang="en"):
+        code_col = "\n".join(args)
+        code, expected = self.converts(code_col)
+        self.assert_highlighted_chr(code, expected, level, lang)
+
+
+
+    # apply rules on code and check if the result is valid
+    def check(self, code, expected, rules):
+        
+        simulator = SimulatorAce(rules)
+        result = simulator.highlight(code)
+
+        valid, ind_error = self.compare(result, expected)
+
+        if not valid:
+
+            code_list     = code.split('\n')
+            expected_list = expected.split('\n')
+            result_list   = result.split('\n')
+
+            line_cpt = 0
+            while ind_error >= len(code_list[line_cpt]):
+                ind_error -= len(code_list[line_cpt]) +1
+                line_cpt += 1
+
+            print("ERROR in this code :")
+            print("In this code :", code_list[line_cpt] )
+            print("We want      :", expected_list[line_cpt] )
+            print("We have      :", result_list[line_cpt] )
+            print("At           :", " " * ind_error + "^")
+
+        self.assertTrue(valid)
+
+
+    # This function allows to compare 2 syntactic colorations one desired
+    # and the other obtained, and manages some special cases
+
+    # Returns a boolean and a value:
+    # - If the 2 highlights are consistent, the boolean will be true and the value unused
+    # - if there is an error in the highlight, the boolean will be false and
+    #   the value will be the number of the 1st character where there is an inconsistency
+    def compare(self, result, expected):
+        if len(result) != len(expected):
+            raise ValueError("The desired highlight and the obtained highlight do not have the same length !")
+
+        cpt = 0
+        for i in range(len(expected)):
+            ch_wanted, ch_result = expected[i], result[i]
+
+            if ch_wanted == ch_result:
+                pass # if they are the same, no problem 
+            elif ch_wanted == " " and ch_result in ['T', 'K', 'C', 'N', 'S']:
+                pass # if it was not fixed at the beginning, everything is tolerated except the pink highlighting
+            else:
+                # in all other cases, an error is returned indicating the number of the problematic character
+                return False, cpt
+
+            cpt += 1
+
+        return True, -1
+
+
+
+    # get rules from json files
+    def get_rules(self, level, lang="en"):
+        os.chdir(os.path.dirname(__file__) +"/..")
```

Comment:
Bad form to `chdir` in general, and definitely in some random utility function here.

Do you think that if some caller code calls `get_rules()`, that they would expect the working directory to be changed out from under them?

The current working directory is like global state: it's scary, anyone can change it and you can't tell what parts of the program changing it might affect. Better not to mess with it at all.

If you can determine the correct directory at all, you can also just open a file directly from that directory without `chdir`ing to it first.

Reply:
You need to combine the two things you were doing:

```py
# Calculate the directory without actually switching to it
root_dir = os.path.dirname(__file__) +"/.."

# Calculate the filename with respect to the correct directory
filename = f'{root_dir}/highlighting/highlighting-trad.json'

# Finally, open the file
with open(filename) as file_regex_trad:
```

Reply:
Done !